### PR TITLE
[NOD-778] Optimize RestoreUTXO

### PR DESCRIPTION
**Pull request by svarogg**
Before looking at the code, please go over the following spreadsheet, and see that you agree with it:
https://docs.google.com/spreadsheets/d/1E8G3mp5y1-yifouwLLXRLueSRfXdDRwRKFieYE07buY/edit?usp=sharing

### CODE REVIEW

**Comment by svarogg** [comment]
> DiffFrom, WithDiff, and WithDiffInPlace became quite far removed from the relatively-straightforward diff algebra we had a while ago.
> Please consider refactoring it to have explicit checks for each case instead of nested if logic.

File: blockdag/utxoset_test.go [hunk 1]
```diff
@@ -460,6 +591,24 @@ func TestUTXODiffRules(t *testing.T) {
 				"Expected: \"%v\", got: \"%v\".", test.name, expectedWithDiffResult, withDiffResult)
 		}
 
+		// Repeat WithDiff check this time using WithDiffInPlace
```
[the remark this answers is not in the document]
Comment (svarogg) [reply blockdag/utxoset_test.go]: I compare both of them to expectedWithDiffResult, so if they are both not equal, at least one of them will fail.

File: blockdag/utxoset_test.go [hunk 2]
```diff
@@ -356,51 +464,76 @@ func TestUTXODiffRules(t *testing.T) {
...
 			},
 			expectedWithDiffResult: &UTXODiff{
 				toAdd:    utxoCollection{outpoint0: utxoEntry1},
-				toRemove: utxoCollection{outpoint0: utxoEntry2},
+				toRemove: utxoCollection{},
 			},
 		},
 		{
-			name: "equal outpoints different blue scores: first in toAdd and second in toRemove in both this and other",
+			name: "empty this, one toRemove in other",
```
Comment (someone235) [inline blockdag/utxoset_test.go]: Please add it to excel

File: blockdag/utxoset.go [hunk 3]
```diff
@@ -331,8 +391,7 @@ func (d *UTXODiff) WithDiff(diff *UTXODiff) (*UTXODiff, error) {
 			// or diff.toRemove.
 			// These are just "updates" to accepted blue score
 			if diffEntry.blockBlueScore != utxoEntry.blockBlueScore &&
-				(d.toRemove.containsWithBlueScore(outpoint, diffEntry.blockBlueScore) ||
-					diff.toRemove.containsWithBlueScore(outpoint, utxoEntry.blockBlueScore)) {
+				diff.toRemove.containsWithBlueScore(outpoint, utxoEntry.blockBlueScore) {
```
[the remark this answers is not in the document]
Comment (svarogg) [reply blockdag/utxoset.go]: The line that I deleted compounds:
1 in d.toAdd
2 in diff.toAdd
2 in d.toRemove

This goes to lines 16 and 20 in the spreadsheet, line 16 is ERROR, while line 20 is satisfied by 
`diff.toRemove.containsWithBlueScore(outpoint, utxoEntry.blockBlueScore)`.

If I was to leave `d.toRemove.containsWithBlueScore(outpoint, diffEntry.blockBlueScore)`, than line 16 would pass as well.

File: blockdag/utxoset.go [hunk 4]
```diff
@@ -353,8 +412,7 @@ func (d *UTXODiff) WithDiff(diff *UTXODiff) (*UTXODiff, error) {
 			// or diff.toAdd.
 			// These are just "updates" to accepted blue score
 			if diffEntry.blockBlueScore != utxoEntry.blockBlueScore &&
-				(d.toAdd.containsWithBlueScore(outpoint, diffEntry.blockBlueScore) ||
-					diff.toAdd.containsWithBlueScore(outpoint, utxoEntry.blockBlueScore)) {
+				d.toAdd.containsWithBlueScore(outpoint, diffEntry.blockBlueScore) {
```
[the remark this answers is not in the document]
Comment (svarogg) [reply blockdag/utxoset.go]: Similar rationale as with previous one. I'll leave this as an exercise for the reader (j/k if you still not sure this is correct - let me know, I'll provide explanation)

File: blockdag/utxoset_test.go (hunk 2 above)
Comment (svarogg) [reply blockdag/utxoset_test.go]: It was in excel named "empty this, first in toRemove in other".
Renamed in code, because this is the naming convention.

**Pull request merged**